Dashboard improvements (8 Oct 2024)

somebody1234 opened the pull request:
### Pull Request Description
- :warning: Follow up to #11219. MUST NOT be merged in before that PR.

- Changes:
  - Add optional overlay to `Popover`s
  - Add and use `useAssetPassiveListener` everywhere to get reactive updates to asset state even outside Asset Panel
  - `setItem` has been removed in favor of simply waiting for invalidations
  - Migrate more `Modal`s to `Popover`s
  - Migrate more inputs to `Form`s
  - Periodically refetch Datalink in Asset Panel
  - Show optimistic state for asset description (adding this because it is trivial to add)
  - Remove usages of `get*` as mutations throughout the entire codebase - replace with `fetchQuery`
- Fixes most of rest of https://github.com/enso-org/cloud-v2/issues/1529
  - (1) :information_source: fixed in #11219
  - (2) :x: backend issue
  - (3) :x: out of scope
  - (4) :x: backend issue
  - (5) :x: out of scope
  - (6) :x: [wontfix]? i think this is intentional, it's not so much slow scrolling and moreso snapped scrolling
  - (7) :x: backend issue
  - (8) :information_source: fixed in #11126
  - (9) :x: out of scope (potentially requires a way to trigger a tooltip on a disabled button)
  - (10) :x: (will check later) Make sure you are not able to open a project opened by another user: cmd + click is not always working.
  - (11) Drag from team space to user space should copy asset
  - (12) :x: (will check later) Drag from user space to team should move (and swap ownership)
  - (13) :information_source: fixed in #11219
  - (14) :information_source: fixed somewhere (?)
  - (15) :information_source: fixed somewhere (?)
  - (16) Show correct (and up-to-date) description for projects
  - (17) :information_source: fixed in #11219
  - (18) :information_source: fixed in #11219
- Fix https://github.com/enso-org/cloud-v2/issues/1535
  - Completely remove optimistic UI for "copy asset"
- Fix https://github.com/enso-org/cloud-v2/issues/1541
  - Make selection brush work again
  - Unintentionally regressed in https://github.com/enso-org/enso/commit/51733ee87677fb47e7a123c98660bc72abe1a9fd#diff-f3e29bffcda342ab6a9dbafc58dde88ce26638eaecda1f17f40ca7e319c90cc8L89

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

somebody1234 committed:
Switch Asset Panel to tab bar

somebody1234 committed:
Fix button variant in `AssetProjectSession`

somebody1234 committed:
Move secret editor to sidebar

somebody1234 committed:
Prevent sidebar from disappearing on Assets Table refresh

somebody1234 committed:
WIP: Place labels above inputs in Datalink dialog

somebody1234 committed:
Disable "Accessed by projects" and "Accessed data" columns

somebody1234 committed:
Unbreak "Manage labels" modal

somebody1234 committed:
WIP: Add checkbox to "manage labels" dialog

somebody1234 committed:
Clean up checkbox list rendering for "manage labels" dialog

somebody1234 committed:
Simplify `Label` component

somebody1234 committed:
Use named imports in `Label`

somebody1234 committed:
Merge branch 'develop' into wip/sb/dashboard-improvements-2024-09-24

somebody1234 committed:
Horizontal scroll for autocomplete

somebody1234 committed:
Fix `Dropdown` styling

somebody1234 committed:
Simplify form cancel buttons

somebody1234 committed:
Rename `formnovalidate` prop to `cancel`

somebody1234 committed:
Allow editing children of JSON schema structs even when they are not yet present on the parent object

somebody1234 committed:
Fix closing asset panel if it is currently temporarily open

somebody1234 committed:
Fix centered settings title

somebody1234 committed:
Use outline variant for form reset forms ("change password" form)

somebody1234 committed:
Use "workspace" icon for project tabs

somebody1234 committed:
Encode URL when copying cloud URLs as path

somebody1234 committed:
Make labels list in Asset Panel (sidebar) horizontal

somebody1234 committed:
Only show billing page to organization admins

somebody1234 committed:
Add support for tooltip saying who is opeing the project (does not work as disabled buttons do not show tooltips)

somebody1234 committed:
Add spotlight

somebody1234 committed:
Fix spotlight

somebody1234 committed:
Treat `isEditingDescription` as true if spotlight is on description edit

somebody1234 committed:
Setting tags on an asset should invalidate "list directory"

somebody1234 committed:
Prettier

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Add `ComboBox`

somebody1234 committed:
Fix `ComboBox`

somebody1234 committed:
Switch `JSONSchemaInput` from `ComboBox` back to `Autocomplete`

somebody1234 committed:
Use `ClearPressResponder` in `TheModal`

somebody1234 committed:
Fix errors?

somebody1234 committed:
Fix double outline for `Autocomplete`

somebody1234 committed:
Simplify `editDescription`, add `useAssetPassiveListener` and `useSetAsset` hooks

somebody1234 committed:
Move `ClearPressResponder` to `Modal`

somebody1234 committed:
Show path even on Cloud backend

somebody1234 committed:
Remove `setItem` from props; use asset passive listener everywhere

somebody1234 committed:
Use named imports

somebody1234 committed:
Show error when failing to load directory

somebody1234 committed:
Display optimistic state for asset description

somebody1234 committed:
Use form for Datalink sidebar input

somebody1234 committed:
Periodically refetch datalink in Asset Panel

somebody1234 committed:
Switch `ResizableContentEditableInput` styles to object instead of destructure

somebody1234 committed:
Fix editing asset description

somebody1234 committed:
Make Label color picker full width in "manage labels" modal

somebody1234 committed:
Add tag after creating in "manage labels" modal

somebody1234 committed:
Truncate `Autocomplete` text

somebody1234 committed:
Fix "new secret" modal

somebody1234 committed:
Autofocus description field when editing

somebody1234 committed:
Fix autocomplete input formatting

somebody1234 committed:
Implement `autoFocus` for `ResizableContentEditableInput`

somebody1234 committed:
Support overlay for `Popover`

somebody1234 committed:
Switch `ManageLabelsModal` from `Modal` to `Popover`

somebody1234 committed:
Add and use `backendQueryOptions`

somebody1234 committed:
Stop using `get*` as mutations

somebody1234 committed:
Switch keyboard shortcut dialog from `Modal` to `Dialog`

somebody1234 committed:
Use named imports

somebody1234 committed:
Update styles according to QA

somebody1234 committed:
Fix crash when editing datalink in right sidebar

somebody1234 committed:
Fixes

somebody1234 committed:
Fix checkbox selected state in "manage labels" modal

somebody1234 committed:
Reactive updates for `useAssetPassiveListener`

somebody1234 committed:
Remove unnecessary `useSetAsset`

somebody1234 committed:
Fix error

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Remove ability to cancel Asset Panel forms

somebody1234 committed:
Merge branch 'wip/sb/dashboard-improvements-2024-09-24' into wip/sb/remove-set-item

somebody1234 committed:
Fix form reset

somebody1234 committed:
Hide "name" input in "upsert secret" modal when editing a secret (not when creating a secret)

somebody1234 committed:
Remove unnecessary parameter to `form.reset()`

somebody1234 committed:
Remove optimistic updates for asset list

somebody1234 committed:
Fix selection brush

somebody1234 committed:
Increase maximum width for "name" column in "members" settings tab

somebody1234 committed:
Show opened by user when opening project is disabled

somebody1234 committed:
Fix issue with inconsistent hook count

somebody1234 committed:
Minor fixes

somebody1234 committed:
Dragging asset from team to user space should copy

somebody1234 committed:
Refactor "copy from team to user space" to a hook; also use hook when doing cut and paste via keyboard shortcut/context menu

somebody1234 committed:
Move paste data into zustand state

somebody1234 committed:
Compute visibility based on zustand `pasteData` state

MrFlashAccount commented on app/dashboard/src/components/AriaComponents/Dialog/Popover.tsx:
```diff
@@ -108,6 +111,15 @@ export function Popover(props: PopoverProps) {
     >
       {(opts) => (
         <dialogStackProvider.DialogStackRegistrar id={dialogId} type="popover">
+          {overlay && (
```
Popovers are not supposed to have overlays

MrFlashAccount commented on app/dashboard/src/components/AriaComponents/Form/components/Reset.tsx:
```diff
@@ -42,14 +42,17 @@ export function Reset(props: ResetProps): React.JSX.Element {
 
   return (
     <ariaComponents.Button
+      onPress={() => {
```
`type="reset"`doesn't work, right?

somebody1234 replied:
yup, i had to remove `type="reset"` - it seems like it uses html `reset` behavior which seems to clear the inputs - i should add a comment

MrFlashAccount replied:
yeah, ideally, we want to use the native api as much as possible, but it seems like it's not always the case though

MrFlashAccount commented on app/dashboard/src/components/AriaComponents/Inputs/ResizableInput/ResizableContentEditableInput.tsx:
```diff
@@ -100,20 +110,26 @@ export const ResizableContentEditableInput = forwardRef(function ResizableConten
     size,
   })
 
+  useEffect(() => {
```
let's replace it with `useAutoFocus`: https://github.com/enso-org/enso/blob/17d75ffdfa764b72b715908bed69ba4c06bd66f5/app/dashboard/src/hooks/autoFocusHooks.ts#L27

Comparing to your snippet, this hook will properly move focus even if we set `autfocus` on dialog close (it'll fight the default focus restore behavior)

MrFlashAccount commented on app/dashboard/src/components/dashboard/AssetRow.tsx:
```diff
@@ -174,12 +179,15 @@ export const AssetRow = React.memo(function AssetRow(props: AssetRowProps) {
+        predicate: (mutation: BackendMutation<'deleteAsset'>) =>
+          mutation.state.variables?.[0] === asset.id,
+      },
+      select: () => true,
```
`select` seems like unnecessary here?

MrFlashAccount commented on app/dashboard/src/hooks/backendHooks.ts:
```diff
@@ -247,3 +270,45 @@ export function useListUserGroupsWithUsers(
+  parentId: DirectoryId | null | undefined,
+  category: Category,
+) {
+  const listDirectoryQuery = useQuery<
```
I think maybe it's worth implementing a hook called `useQueryCache`, which will select the query by partial matching the key or by a predicate?
similar to `useQueryClient().getQueryCache().find()`: https://tanstack.com/query/latest/docs/reference/QueryCache#querycachefind but with a subscription.

Advantage is in the fact that we don't need to know all the queryKey parameters to find the asset

somebody1234 replied:
true. but in this case i'm a little bit worried about performance so i would probably prefer `useQuery` still in this case - we may need to eventually end up calling this hook for every row

MrFlashAccount replied:
I’m curious, what are you specifically worried about with this implementation?

somebody1234 replied:
hmm... i guess it doesn't hurt to switch to `find`... not sure how their filtering logic works actually

somebody1234 replied:
ok on second thought: i think the hook will be easy to implement, but unfortunately it is not an option here as all our queryKey parameters are important here (since categories affect queryKey params here)

MrFlashAccount replied:
But if found an asset with the same id, does it mean that we found the correct one?

somebody1234 replied:
true... but these are filters for the parent directory, we probably want to avoid looking through every directory listing just because we don't have the info. so what i think rn is:

- yes fuzzy matching for an asset *might* be something we want in the future
- however we currently have all the info we need in all the places we need this query
- and fuzzier matches means more data matched means more lists we need to scan through to find our asset

(IMO the ideal solution is to create like, derived queries for individual assets so we can avoid needing the parent id at all - just the current asset's id and nothing else)

somebody1234 committed:
WIP: Fix "asset not found" invariant error

somebody1234 committed:
Return dummy results for `Users` and `Teams`

somebody1234 committed:
Fix directory queries

somebody1234 committed:
Show "shared with" in sidebar only for "Enterprise" organizations

somebody1234 committed:
Address CR

somebody1234 committed:
Remove `Popover overlay`

somebody1234 committed:
Add `useCachedQueries` hook

somebody1234 committed:
Fixes

somebody1234 committed:
Merge branch 'develop' into wip/sb/remove-set-item

somebody1234 commented:
assigning @MrFlashAccount for QA (not super urgent of course!)

somebody1234 committed:
Merge branch 'develop' into wip/sb/remove-set-item

somebody1234 committed:
Merge branch 'develop' into wip/sb/remove-set-item

PabloBuchu commented:
1. Something broke with datalink modal
<img width="682" alt="Screenshot 2024-10-11 at 12 15 37" src="https://github.com/user-attachments/assets/c16b7825-3315-4bc1-9275-31e3ea4fda1d">

2. Box selection is only working for bottom assets... because it is limited to the asset table I can't select 3 top ones or just the ones in the middle... should we fix / improve this as separate PR?
3. Can't open a local project... 
<img width="810" alt="Screenshot 2024-10-11 at 12 28 31" src="https://github.com/user-attachments/assets/f6319246-80d0-421d-9911-1c910c41f118">

somebody1234 committed:
Increase margin of selection brush

somebody1234 commented:
> [box selection] is limited to the asset table

this one is an easy fix, because actually we had such behavior before

somebody1234 committed:
Fix styles for datalink modal

somebody1234 committed:
Fix opening projects

somebody1234 committed:
Fix error

somebody1234 committed:
Use new paste validity detection in `AssetContextMenu`

somebody1234 commented:
@MrFlashAccount i guess this one is awaiting re-CR?

somebody1234 committed:
Merge branch 'develop' into wip/sb/remove-set-item

somebody1234 committed:
Prettier

somebody1234 committed:
Merge branch 'develop' into wip/sb/remove-set-item

somebody1234 committed:
`useCutAndPaste` should fall back to `useTransferBetweenCategories`

FIXME: `useTransferBetweenCategories` should use `copy` instead of `move` when transferring from the Team Space to the User Space

somebody1234 committed:
woops

somebody1234 committed:
Add command to E2E test dashboard

somebody1234 committed:
Stop specialcasing trash category in `AssetRow` event handling

somebody1234 committed:
Fix lint error

somebody1234 committed:
Fix invariant violation when un-deleting asset

somebody1234 committed:
More fixes for invalid asset panel state

somebody1234 committed:
Clear asset panel state on drop

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Add `useBackendMutationState` hook

somebody1234 committed:
Fix spurious re-selection after deselection

somebody1234 committed:
Prettier

somebody1234 committed:
Fix errors
